### DISCUSSION TITLE
working example for automatic admin + user creation + configuration to call a webhook and store uploaded files in an S3 bucket?

### DISCUSSION THREAD

**Discussion by maspotts** [drakkan/sftpgo#1487] [Q&A]
Hi: I have the following:

  1. an SSL cert + key + ca_bundle;
  2. an SSH private key
  3. an S3 bucket and associated access key + secret;
  4. a function (python for now) which generates a filename by generating a UUID and timestamp and concatenating them.
  
I want to be able to deploy sftgo as a service, via Docker, so that it can be launched as a microservice without human interaction and each time it launches it should be able to  receive uploads via HTTP(S), FTP(S) (using my cert + key) and SFTP (using my SSH private key), call my filename generation web service, and store the uploaded file in my S3 bucket (under the generated filename).

In my initial experimentation I've been unable to get the SSL cert to work, and I've been unable to get sftpgo to automatically generated an admin user via the command-line.  Rather than ask for help step-by-step, I'm hoping someone might be able to reply with some example JSON and shell code that I could adapt to my use case?  I think I need:

  1. a config.json correctly configured to specify HTTP, FTP and SFTP and to apply the cert, key and SSH key, and specify automatic creation of an admin user via environment variables at launch time;
  2. a bash/sh command-line that specifies an admin username and password via the environment and launches sftpgo with them.
  3. an add-user.json correctly configured to add a non-admin user, correctly configured to call my name generation web server and store files in my S3 bucket.
  
If that's not too difficult then I will hugely appreciate the help!  If it's too much to ask, then I totally understand and will stop pestering this forum for help :)

Thanks!

**Comment by drakkan** [comment]
What you ask is doable even without an external hook, the only thing missing is adding a placeholder to generate a UUID in SFTPGo actions which is our private list of things to check feasibility for.

I suggest mounting a volume with your certificates, keys, and a pre-populated database (maybe just a dump for the memory provider). I have no time to provide further details, sorry. Maybe someone else can help.

**Reply by drakkan** [reply]
https://github.com/drakkan/sftpgo/commit/00ce4e46851d97bcda20c348aa7ebd4cbad95288

so an external hook is not strictly required

**Comment by maspotts** [comment]
Very cool!  Could you also make an SHA2 (or similar) hash of the file contents available (again without an external hook)?  That's my only remaining need (I would ideally like to save the files as "EPOCHTIME+UUID+SHA2".)

**Comment by drakkan** [comment]
> Very cool! Could you also make an SHA2 (or similar) hash of the file contents available (again without an external hook)? That's my only remaining need (I would ideally like to save the files as "EPOCHTIME+UUID+SHA2".)

Not planned for now, sorry. This requires design decision and probably a new action. Feature requests like this are only accepted from Enterprise plan subsribers. Good luck with your project
